Add mermaids for file content encryption. (#30)
encryption-alliance/unified-vault-format: file name encryption/AES-SIV-512-B64URL.md

   decision -->|n| csprng32
   decision -->|y| initialSeed
   latestSeed --> directorySeed
   csprng32 --> dirId
   csprng32{{"csprng(32)"}}
   initialSeed --> directorySeed
   initialSeed -->|seed:| kdfRootDirId
   kdfRootDirId{{"kdf(seed,32,'rootDirId')"}}
   kdfRootDirId --> dirId
   directorySeed -->|seed:| kdfSiv
   kdfSiv{{"kdf(seed,64,'siv')"}}
   kdfSiv --> sivKey
   directorySeed -->|seed:| kdfHmac
   kdfHmac{{"kdf(seed,32,'hmac')"}}
   kdfHmac --> hmacKey
   hmacKey -->|key:| hmacSha256
   hmacSha256{{hmacSha256}}
   hmacSha256 --> dirIdHash
   dirIdHash --> truncate
   truncate{{"_[0..20]"}}
   truncate --> base32
   base32{{base32}}
   base32 --> dirIdString
   dirIdString --> head
   head{{"_[0..2]"}}
   dirIdString --> tail
   tail{{"_[2..32]"}}
   head -->|$0:| pattern
   tail -->|$1:| pattern
   pattern{{"'d/'$0'/'$1"}}
   pattern --> dirPath
   sivKey -->|sivKey:| aesSiv
   aesSiv{{aesSiv}}
   parentDirId -->|ad:| aesSiv
   clearTextName -->|secret:| aesSiv
   aesSiv --> base64Url
   base64Url{{base64url}}
   base64Url --> join
   uvf --> join
   uvf["'.uvf'"]
   join{{"join"}}
   join --> ciphertextName
   directorySeed --> concat
   dirId --> concat

>    concat{{concat}}

Reviewer: This is incorrect. We do not concatenate seed and dirId. Instead we create a normal encrypted file whose sole content is the `dirId`. The seed is used in the file header (as correctly indicated by `directorySeed -->|seed:| fileContentEncryption`)

Author: I was mislead by

> Every directory requires certain metadata that affects the file name encryption of its direct children:
> 
> * the seed used to derive keys
> * the directory ID

into thinking that both go in to the body. However, it's actually clear later on:

> The cleartext content of `dir.uvf` is the 32 byte dirId.

Author: Done: 5ca480306b88a46cf711b845e89a2d5bdd2dcdb9

   concat -->|cleartextBlocks:| fileContentEncryption
   fileContentEncryption{{file content encryption}}
   directorySeed -->|seed:| fileContentEncryption
   fileContentEncryption --> dirUvf
   dirUvf["directory metadata dir.uvf content"]
```

> 

Reviewer: I like the idea, but let's break this into two charts to make it more digestible:

1. decide which dirId and seed is used
2. derive dir path and encrypt node names

Author: Done: 92dbd884e51ba0dac60d8b024af0f6ff702a6bed

#### Read content of `/File.txt`:

1. decrypt file `d/BZ/R4VZSS5PEF7TU3PMFIMON5GJRNBDWA/5TyvCyF255sRtfrIv83ucADQ.uvf`

## Overview

### Derivation of Directory Seed and Directory ID
```mermaid

> flowchart TD
>     decision{root?}
>     decision -->|n| latestSeed
>     decision -->|n| csprng32
>     decision -->|y| initialSeed
>     latestSeed --> directorySeed
>     csprng32 --> dirId
>     csprng32{{"csprng(32)"}}
>     initialSeed --> directorySeed
>     initialSeed -->|secret:| kdfRootDirId
>     kdfRootDirId{{"kdf(secret,32,'rootDirId')"}}
>     kdfRootDirId --> dirId
>     subgraph Outputs
>         directorySeed
>         dirId
>     end

Reviewer: how about this?

```suggestion
---
title: Choosing a Seed and Directory ID
---
flowchart TD
    subgraph "Which Seed?"
        isRoot1{is root dir?}
        isRoot1 -->|y| initialSeed
        isRoot1 -->|n| isExistingDir1

        isExistingDir1{is existing dir?}
        isExistingDir1 -->|y| readSeed
        isExistingDir1 -->|n| latestSeed

        readSeed{{"read seed from `dir.uvf`"}}
        readSeed --> dirFileSeed[seed from dir.uvf]
    end

    subgraph "Which dirId?"
        isRoot2{is root dir?}
        isRoot2 -->|n| isExistingDir2
        isRoot2 -->|y| kdfRootDirId

        isExistingDir2{is existing dir?}
        isExistingDir2 -->|y| readDirId
        isExistingDir2 -->|n| csprng32
        
        csprng32 --> randomDirId[random dirId]
        csprng32{{"csprng(32)"}}

        initialSeed -->|secret:| kdfRootDirId    
        kdfRootDirId{{"kdf(secret,32,'rootDirId')"}}
        kdfRootDirId --> rootDirId[root dirId]

        readDirId{{"read dirId from `dir.uvf`"}}
        readDirId --> dirFileId[dirId from dir.uvf]
    end
```

```

### Mapping Directory IDs to Paths, Encryption of Directory Names and Directory Metadata
```mermaid
flowchart TD
   directorySeed -->|secret:| kdfSiv
   kdfSiv{{"kdf(secret,64,'siv')"}}
   kdfSiv --> sivKey
   directorySeed -->|secret:| kdfHmac
   kdfHmac{{"kdf(secret,32,'hmac')"}}
   kdfHmac --> hmacKey
   hmacKey -->|key:| hmacSha256
   hmacSha256{{hmacSha256}}
   hmacSha256 --> dirIdHash
   dirIdHash --> truncate
   truncate{{"_[0..20]"}}
   truncate --> base32
   base32{{base32}}
   base32 --> dirIdString
   dirIdString --> head
   head{{"_[0..2]"}}
   dirIdString --> tail
   tail{{"_[2..32]"}}
   head -->|$0:| pattern
   tail -->|$1:| pattern

>    pattern{{"'d/'$0'/'$1"}}

Reviewer: let's assume a language that can deal with a simpler pattern string 😉

```suggestion
   pattern{{"d/$0/$1"}}
```

   pattern --> dirPath
   sivKey -->|sivKey:| aesSiv
   aesSiv{{aesSiv}}
   parentDirId -->|ad:| aesSiv
   clearTextName -->|secret:| aesSiv
   aesSiv --> base64Url
   base64Url{{base64url}}

>    base64Url --> join
>    uvf --> join
>    uvf["'.uvf'"]
>    join{{"join"}}
>    join --> ciphertextName

Reviewer: graph direction is odd (looks like `join('.uvf', base64)`). let's simplify:

```suggestion
   base64Url --> appendFileExt
   appendFileExt{{"append '.uvf'"}}
   appendFileExt --> ciphertextName
```
